runners/direct-java/src/main/java/org/apache/beam/runners/direct/StepTransformResult.java

```diff
@@ -57,6 +59,13 @@
   @Override
   public abstract TimerUpdate getTimerUpdate();
 
+  @Override
+  public boolean producedOutput() {
+    return !Iterables.isEmpty(getOutputBundles()) || isProducedAdditionalOutput();
+  }
+
+  abstract boolean isProducedAdditionalOutput();
```

Comment:
Javadoc.

Also, the mixing of `producedOutput` and `isProducedAdditionalOutput` is odd. Either both start with `is` or neither start with `is`.